File: rust_luogu/src/bin/p1525.rs
//@author    Allergy
//@email     Allergy527@gmail.com
//@workspace Prectice_Contest/p1525.rs
//@data      2024/10/22 17:56:39
fn cin() -> String {
    let mut input = String::new();
    std::io::stdin().read_line(&mut input).unwrap();
    input.trim().to_string()
}
fn main() {
    let t = 1; //let t = cin().parse::<i32>().unwrap();
    let k = (1..=4).fold(1, |ans, x| ans + (1..=x).product::<i32>());
    println!("{}", k);
    (0..t).for_each(|_| solve());
}
struct UFset {
    fa: Vec<usize>,
    rel: Vec<usize>,
}
impl UFset {
    fn new(n: usize) -> Self {
        Self {
            fa: (0..=n).collect::<Vec<_>>(),
            rel: vec![0; n + 1],
        }
    }
    fn find(&mut self, x: usize) -> usize {
        if x != self.fa[x] {
            let tmp = self.fa[x];
            self.fa[x] = self.find(self.fa[x]);
            self.rel[x] ^= self.rel[tmp];
        }
        self.fa[x]
    }
    fn merge(&mut self, x: usize, y: usize, fx: usize, fy: usize) {
        self.fa[fy] = fx;
        self.rel[fy] = self.rel[x] ^ self.rel[y] ^ 1;
    }
}

fn solve() {
    let (n, m) = cin()
        .split_whitespace()
        .fold((0, 0), |x, y| (x.1, y.parse::<usize>().unwrap()));
    let mut qwq = vec![(0, 0, 0); m];
    for x in &mut qwq {
        *x = cin()
            .split_whitespace()
            .fold((0, 0, 0), |x, y| (x.1, x.2, y.parse::<usize>().unwrap()));
    }
    qwq.sort_unstable_by(|a, b| b.2.cmp(&a.2));
    let mut uf = UFset::new(n);
    for &(a, b, w) in &qwq {
        let (fa, fb) = (uf.find(a), uf.find(b));
        if fa == fb {
            if uf.rel[a] == uf.rel[b] {
                println!("{}", w);
                return;
            }
        } else {
            uf.merge(a, b, fa, fb);
        }
    }
    println!("0")
}

File: cpp/luogu/old/day2/p1525.cpp
// Copyright 2024 Allergy
// email     Allergy527@gmail.com
// workspace luogu\day2\p1525.cpp
// date      2024/09/23 21:19:38
#include <bits/stdc++.h>
// #define int int64_t

using namespace std;  // NOLINT
using ll = int64_t;
int inf = 0x3f3f3f3f;
ll ll_inf = 0x3f3f3f3f3f3f3f3f;

struct Node {
    int x, y, w;
    bool operator<(const Node &rhs) const {
        return w > rhs.w;
    }
};

struct UFSet {
    vector<int> fa, rel;
    UFSet(int n) : fa(n + 1), rel(n + 1) {  // NOLINT
        iota(fa.begin(), fa.end(), 0);
    }
    int find(int x) {
        if (fa[x] == x) return x;
        int temp = find(fa[x]);
        rel[x] = rel[x] ^ rel[fa[x]];
        return fa[x] = temp;
    }
    void merge(int x, int y, int rootx, int rooty) {
        int relation = (rel[x] + rel[y] + 1) % 2;
        fa[rooty] = rootx;
        rel[rooty] = relation;
    }
};

void solve() {
    int n, m;
    cin >> n >> m;
    vector<Node> qwq(m);
    for (auto &[x, y, w] : qwq) cin >> x >> y >> w;
    sort(qwq.begin(), qwq.end());
    UFSet qvq(n);
    int ans = 0;
    for (int i = 0; i < m; ++i) {
        int x = qwq[i].x, y = qwq[i].y, w = qwq[i].w;
        int rootx = qvq.find(x), rooty = qvq.find(y);
        if (rootx != rooty) {
            qvq.merge(x, y, rootx, rooty);
        } else {
            int relation = qvq.rel[x] ^ qvq.rel[y];
            if (relation == 0) {
                ans = w;
                break;
            }
        }
    }
    cout << ans << '\n';
}

signed main() {
    ios::sync_with_stdio(false);
    int t = 1;
    // cin >> t;
    while (t--) solve();
    cin >> t;
    return 0;
}
